variables/parse.go

```diff
@@ -88,20 +83,36 @@ func parseArg(arg string, pncvf ProcessNonConstantVariableFunc, current *Collect
 	return name, v, nil
 }
 
-func parseArgValue(name string, value string, pncvf ProcessNonConstantVariableFunc) (string, error) {
-	if pncvf == nil {
-		return value, nil
-	}
-	if strings.HasPrefix(value, "$(") {
-		// Variable build arg - resolve value.
-		var err error
-		value, _, err = pncvf(name, value)
-		if err != nil {
-			return "", err
+// ContainsShell returns true for strings containing $(
+// except for cases where escaped: e.g. \$(
+// or cases with singlquotes: '$(...'
+func ContainsShell(s string) bool {
+	var escaped bool
+	var singlequoted bool
+	var last rune
+	for _, c := range s {
+		//fmt.Printf("got %s\n", string(c))
+		if escaped {
+			escaped = false
+			last = 0
+			continue
 		}
+		if c == '\\' {
+			escaped = true
+			last = 0
+			continue
+		}
+		if c == '\'' {
+			singlequoted = !singlequoted
+			last = 0
+			continue
+		}
+		if last == '$' && c == '(' && !singlequoted {
+			return true
+		}
+		last = c
 	}
-
-	return value, nil
+	return false
```

Comment:
I would try to reuse this code: https://github.com/moby/buildkit/blob/master/frontend/dockerfile/shell/lex.go#L13. It would require modifications, to look for `$(...)` and not just for `$ABC` and `${...}`, but I think it already covers a lot of edge cases and it's battle-tested.

Reply:
The tricky part is this code would have to run in a container attached to the current llb.State and not within buildkitd.

Reply:
Yeah.. we'd have to use a function as the value to replace the variable with.

Maybe it's overkill - no strong opinion either way. Maybe it's better to have our own code for this.